atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Sets.SetView;
+import com.google.common.io.BaseEncoding;
+import com.google.common.primitives.UnsignedBytes;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.keyvalue.api.InsufficientConsistencyException;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.common.base.FunctionCheckedException;
+import com.palantir.common.concurrent.PTExecutors;
+
+/**
+ * Feature breakdown:
+ *   - Pooling
+ *   - Token Aware Mapping / Query Routing / Data partitioning
+ *   - Retriable Queries
+ *   - Pool member error tracking / blacklisting*
+ *   - Pool refreshing
+ *   - Pool node autodiscovery
+ *   - Pool member health checking*
+ *
+ *   *entirely new features
+ *
+ *   By our old system, this would be a RefreshingRetriableTokenAwareHealthCheckingManyHostCassandraClientPoolingContainerManager;
+ *   ... this is one of the reasons why there is a new system.
+ **/
+public class CassandraClientPool {
+    private static final Logger log = LoggerFactory.getLogger(CassandraClientPool.class);
+    private static long BANNED_HOST_BACKOFF_TIME_MS = TimeUnit.MINUTES.toMillis(5);
+    private static int MAX_TRIES = 3;
+
+    RangeMap<LightweightOPPToken, List<InetSocketAddress>> tokenMap = ImmutableRangeMap.of();
+    Map<InetSocketAddress, Long> blacklistedHosts = Maps.newHashMap();
+    Map<InetSocketAddress, CassandraClientPoolingContainer> currentPools = Maps.newHashMap();
+    CassandraKeyValueServiceConfig config;
+    ScheduledThreadPoolExecutor refreshDaemon;
+
+
+    private class LightweightOPPToken implements Comparable<LightweightOPPToken> {
+        byte[] bytes;
+
+        public LightweightOPPToken(byte[] bytes) {
+            this.bytes = bytes;
+        }
+
+        @Override
+        public int compareTo(LightweightOPPToken other) {
+            return UnsignedBytes.lexicographicalComparator().compare(this.bytes, other.bytes);
+        }
+    }
+
+    public CassandraClientPool(CassandraKeyValueServiceConfig config) {
+        this.config = config;
+        refreshDaemon = PTExecutors.newScheduledThreadPool(1, new ThreadFactoryBuilder().setDaemon(true).setNameFormat("CassandraClientPoolingContainerRefreshDaemon-%d").build());
+        refreshDaemon.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    refreshPool();
+                } catch (Throwable t) {
+                    log.error("Failed to refresh Cassandra KVS pool. Extended periods of being unable to refresh will cause perf degradation.", t);
+                }
+            }
+        }, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, TimeUnit.SECONDS);
```

Comment:
Sure. Might punt on this to a later commit, along with also making timeout-period-to-attempt-to-add-blacklisted-hosts-back-into-the-fold period.